File: src/ios/VLCPlayer.m
#import "VLCPlayer.h"

@interface VLCPlayer ()
@property (nonatomic, strong) VLCMediaPlayer *player;
@property (nonatomic, strong) UIViewController *containerVC;
@property (nonatomic, strong) UIView *videoView;
@end

@implementation VLCPlayer

- (void)init:(CDVInvokedUrlCommand*)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.containerVC = [UIViewController new];
        self.containerVC.view.backgroundColor = UIColor.blackColor;

        self.videoView = [[UIView alloc] initWithFrame:[UIScreen mainScreen].bounds];
        self.videoView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [self.containerVC.view addSubview:self.videoView];

        CDVPluginResult *res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:res callbackId:command.callbackId];
    });
}

- (void)play:(CDVInvokedUrlCommand*)command {
    NSString *urlStr = ([command.arguments count] > 0 ? command.arguments[0] : nil);
    if (urlStr.length == 0) {
        [self sendError:@"Invalid URL" cb:command.callbackId]; return;
    }
    NSDictionary *opts = ([command.arguments count] > 1 ? command.arguments[1] : @{});
    NSNumber *nc = opts[@"networkCaching"] ?: @(1000);

    dispatch_async(dispatch_get_main_queue(), ^{
        if (!self.containerVC) {
            [self init:command]; // prepares container
        }
        if (self.containerVC.presentingViewController == nil && self.viewController) {
            [self.viewController presentViewController:self.containerVC animated:YES completion:nil];
        }

        // Build VLC player
        NSArray<NSString*> *args = @[[NSString stringWithFormat:@"--network-caching=%@", nc]];
        self.player = [[VLCMediaPlayer alloc] initWithOptions:args];
        self.player.delegate = self;
        self.player.drawable = self.videoView;

        NSURL *url = [NSURL URLWithString:urlStr];
        self.player.media = [VLCMedia mediaWithURL:url];

        [self.player play];

        CDVPluginResult *ok = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:ok callbackId:command.callbackId];
    });
}

- (void)pause:(CDVInvokedUrlCommand*)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.player pause];
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK]
                                    callbackId:command.callbackId];
    });
}

- (void)resume:(CDVInvokedUrlCommand*)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.player play];
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK]
                                    callbackId:command.callbackId];
    });
}

- (void)seek:(CDVInvokedUrlCommand*)command {
    NSNumber *ms = ([command.arguments count] ? command.arguments[0] : @(0));
    dispatch_async(dispatch_get_main_queue(), ^{
        VLCTime *length = self.player.media.length;
        if (length.intValue > 0) {
            float pos = ms.doubleValue / length.doubleValue; // 0..1
            self.player.position = pos;
        }
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK]
                                    callbackId:command.callbackId];
    });
}

- (void)position:(CDVInvokedUrlCommand*)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        int timeMs = self.player.time.intValue ?: 0;
        int lenMs  = self.player.media.length.intValue ?: 0;
        NSDictionary *payload = @{ @"time": @(timeMs), @"length": @(lenMs) };
        CDVPluginResult *res = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:payload];
        [self.commandDelegate sendPluginResult:res callbackId:command.callbackId];
    });
}

- (void)stop:(CDVInvokedUrlCommand*)command {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.player stop];
        self.player = nil;
        [self.containerVC dismissViewControllerAnimated:YES completion:nil];
        self.containerVC = nil;
        self.videoView = nil;
        [self.commandDelegate sendPluginResult:[CDVPluginResult resultWithStatus:CDVCommandStatus_OK]
                                    callbackId:command.callbackId];
    });
}

- (void)sendError:(NSString*)msg cb:(NSString*)cb {
    CDVPluginResult *err = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:msg];
    [self.commandDelegate sendPluginResult:err callbackId:cb];
}

@end

File: src/ios/VLCPlayer.swift
import Foundation
import MobileVLCKit

@objc(VLCPlayer) class VLCPlayer: CDVPlugin, VLCMediaPlayerDelegate {
    private var player: VLCMediaPlayer?
    private var containerVC: UIViewController?
    private var videoView: UIView?

    @objc(init:)
    func `init`(_ command: CDVInvokedUrlCommand) {
        // Prepare a simple fullscreen container
        DispatchQueue.main.async {
            let vc = UIViewController()
            vc.view.backgroundColor = .black

            let v = UIView(frame: UIScreen.main.bounds)
            v.autoresizingMask = [.flexibleWidth, .flexibleHeight]
            vc.view.addSubview(v)

            self.videoView = v
            self.containerVC = vc

            let result = CDVPluginResult(status: CDVCommandStatus_OK)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }

    @objc(play:)
    func play(_ command: CDVInvokedUrlCommand) {
        guard let urlString = command.argument(at: 0) as? String,
              let url = URL(string: urlString) else {
            sendError("Invalid URL", command.callbackId); return
        }
        let optsDict = (command.argument(at: 1) as? [String: Any]) ?? [:]
        let caching = (optsDict["networkCaching"] as? Int) ?? 1000

        DispatchQueue.main.async {
            if self.containerVC == nil { self.init(CDVInvokedUrlCommand()) }

            // Present our container if not on screen
            if let root = self.viewController, let vc = self.containerVC, vc.presentingViewController == nil {
                root.present(vc, animated: true, completion: nil)
            }

            // Build VLC player
            let args = ["--network-caching=\(caching)"]
            let p = VLCMediaPlayer(options: args)
            p.delegate = self
            p.drawable = self.videoView

            let media = VLCMedia(url: url)
            p.media = media

            self.player = p
            p.play()

            let ok = CDVPluginResult(status: CDVCommandStatus_OK)
            self.commandDelegate.send(ok, callbackId: command.callbackId)
        }
    }

    @objc(pause:)
    func pause(_ command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            self.player?.pause()
            self.commandDelegate.send(CDVPluginResult(status: .ok), callbackId: command.callbackId)
        }
    }

    @objc(resume:)
    func resume(_ command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            self.player?.play()
            self.commandDelegate.send(CDVPluginResult(status: .ok), callbackId: command.callbackId)
        }
    }

    @objc(seek:)
    func seek(_ command: CDVInvokedUrlCommand) {
        let ms = (command.argument(at: 0) as? NSNumber)?.doubleValue ?? 0
        DispatchQueue.main.async {
            if let length = self.player?.media?.length, length.intValue > 0 {
                // VLC seek uses position 0..1
                let pos = Float(ms / length.doubleValue)
                self.player?.position = pos
            }
            self.commandDelegate.send(CDVPluginResult(status: .ok), callbackId: command.callbackId)
        }
    }

    @objc(position:)
    func position(_ command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            let timeMs = self.player?.time.intValue ?? 0
            let lenMs  = self.player?.media?.length.intValue ?? 0
            let payload: [String: Any] = ["time": timeMs, "length": lenMs]
            let res = CDVPluginResult(status: .ok, messageAs: payload)
            self.commandDelegate.send(res, callbackId: command.callbackId)
        }
    }

    @objc(stop:)
    func stop(_ command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            self.player?.stop()
            self.player = nil
            // Dismiss container
            self.containerVC?.dismiss(animated: true, completion: nil)
            self.containerVC = nil
            self.videoView = nil
            self.commandDelegate.send(CDVPluginResult(status: .ok), callbackId: command.callbackId)
        }
    }

    // Helpers
    private func sendError(_ msg: String, _ cb: String?) {
        let res = CDVPluginResult(status: .error, messageAs: msg)
        self.commandDelegate.send(res, callbackId: cb)
    }
}
